src/dfx/src/lib/operations/canister/install_canister.rs

```diff
@@ -136,7 +140,47 @@ pub async fn install_canister(
         )
         .await?;
     }
-
+    let mut times = 0;
+    loop {
+        match agent
+            .read_state_canister_info(canister_id, "module_hash", false)
+            .await
+        {
+            Ok(reported_hash) => {
+                if reported_hash == new_hash {
+                    break;
+                } else if installed_module_hash
+                    .as_deref()
+                    .map_or(true, |old_hash| old_hash == reported_hash)
+                {
+                    times += 1;
+                    if times % 4 == 0 {
+                        info!(
+                            env.get_logger(),
+                            "Waiting for module change to be reflected in system state tree..."
+                        )
+                    }
+                    sleep(Duration::from_millis(500)).await;
+                } else {
+                    bail!("The reported module hash is neither the existing module or the new one. \
+                        It has likely been modified while this command is running. \
+                        The state of the canister is unknown. \
+                        For this reason, no post-installation tasks have been run, including asset uploads.")
```

Comment:
Can we display the three module hashed involved here?  That will make it easier for the developer to follow up with `dfx canister info` and troubleshoot.